### PR TITLE
gh-246: replace `numpy.ndarray` with `numpy.typing.NDArray`

### LINKED ISSUE #246: Use `numpy.typing` instead of `numpy.ndarray`
### Is Your Feature Request Related to a Problem? Please Describe

This allows greater control of the type of a specific NumPy array https://numpy.org/doc/stable/reference/typing.html

### Describe the Solution You'd Like

_No response_

### Describe Alternatives You've Considered

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by paddyroddy**
Closes #246. Use `numpy.typing.NDArray` as the first step towards #280. With `NDArray` we can use types, i.e. `NDArray[np.float64]`, which I'll save for future work (#306).

I have also gone throughout and made sure anything like `ArrayLike` is using the alias `import numpy.typing as npt`. Imports are generally in a `if TYPE_CHECKING` block where possible.

We are using `numpy.testing` in some places. This was being aliased as `npt`. To avoid confusion, I have removed this alias and left them as an explicit import.

### CODE REVIEW

File: glass/fields.py [hunk 1]
```diff
@@ -29,25 +29,24 @@
...
-Size = Union[None, int, Tuple[int, ...]]
-Iternorm = Tuple[Optional[int], Array, Array]
-ClTransform = Union[str, Callable[[Array], Array]]
-Cls = Sequence[Union[Array, Sequence[float]]]
-Alms = np.ndarray
+Size = None | int | tuple[int, ...]
+Iternorm = tuple[int | None, npt.NDArray, npt.NDArray]
+ClTransform = str | Callable[[npt.NDArray], npt.NDArray]
+Cls = Sequence[npt.NDArray | Sequence[float]]
+Alms = npt.NDArray
```
Comment (Saransh-cpp) [inline glass/fields.py:44]: I have tried this before and the newer syntax does not work when you are making your own custom types.

Comment (paddyroddy) [reply glass/fields.py]: Well let's see. I got an error, but that looked to be from it being `None | <x>` rather `<x> | None`.

Comment (paddyroddy) [reply glass/fields.py]: Yeah, it was that. Have restored `Optional`, `Tuple`, `Union` here

**Review by Saransh-cpp — APPROVED** [review]
This looks good, thanks, @paddyroddy!

**Comment by paddyroddy** [comment]
This is ready, but blocked by #307.

**Comment by paddyroddy** [comment]
Actually thinking about it, the test failure is related to the notebooks - which aren't being touched. So actually this is okay to merge.

**Pull request merged**